kratos/solving_strategies/convergencecriterias/mixed_generic_criteria.h

```diff
@@ -0,0 +1,451 @@
+     * @param rDofSet Reference to the container of the problem's degrees of freedom (stored by the BuilderAndSolver)
+     * @param rDx Vector of results (variations on nodal variables)
+     * @return std::tuple<std::vector<TDataType>, std::vector<TDataType>> Tuple containing the absolute and relative convergence values
+     */
+    std::tuple<std::vector<TDataType>, std::vector<TDataType>> CalculateConvergenceNorms(
+        const ModelPart& rModelPart,
+        const DofsArrayType& rDofSet,
+        const TSystemVectorType& rDx)
+    {
+        // Initialize
+        std::vector<int> dofs_count(mVariableSize, 0);
+        std::vector<TDataType> solution_norms_vector(mVariableSize, 0.0);
+        std::vector<TDataType> increase_norms_vector(mVariableSize, 0.0);
+
+        // Accumulate the norm values
+        GetNormValues(rModelPart, rDofSet, rDx, dofs_count, solution_norms_vector, increase_norms_vector);
+
+        // Synchronize the norm values
+        const auto& r_data_comm = rModelPart.GetCommunicator().GetDataCommunicator();
+        auto global_solution_norms_vector = r_data_comm.SumAll(solution_norms_vector);
+        auto global_increase_norms_vector = r_data_comm.SumAll(increase_norms_vector);
+        auto global_dofs_count = r_data_comm.SumAll(dofs_count);
+
+        // Check division by zero in global solution norms
+        const double zero_tol = 1.0e-12;
+        for(int i = 0; i < mVariableSize; i++) {
+            if (global_solution_norms_vector[i] < zero_tol) {
+                global_solution_norms_vector[i] = 1.0;
+            }
+        }
+
+        // Calculate the norm values
+        std::vector<TDataType> var_ratio(mVariableSize, 0.0);
+        std::vector<TDataType> var_abs(mVariableSize, 0.0);
+        for(int i = 0; i < mVariableSize; i++) {
+            var_ratio[i] = std::sqrt(global_increase_norms_vector[i] / global_solution_norms_vector[i]);
+            var_abs[i] = std::sqrt(global_increase_norms_vector[i]) / static_cast<TDataType>(global_dofs_count[i]);
+        }
+
+        // Output the ratio and absolute norms as a tuple
+        return std::make_tuple(var_ratio, var_abs);
+    }
+
+    /**
+     * @brief Method to output the convergence status
+     * This method prints the convergence status to the screen for each one of the checked variables
+     * @param rConvergenceNorms Tuple containing the absolute and relative convergence values
+     */
+    virtual void OutputConvergenceStatus(
+        const std::tuple<std::vector<TDataType>,
+        std::vector<TDataType>>& rConvergenceNorms)
+    {
+        const auto& var_ratio = std::get<0>(rConvergenceNorms);
+        const auto& var_abs = std::get<1>(rConvergenceNorms);
+
+        if (this->GetEchoLevel() > 0) {
+            std::ostringstream stringbuf;
+            stringbuf << "CONVERGENCE CHECK:\n";
+            for(int i = 0; i < mVariableSize; i++) {
+                const auto r_var_data = mVariableDataVector[i];
```

Comment:
Probably a protected method to access `mVariableDataVector`, just like what we have for `mVariableSize`?

Reply:
I created a set of protected methods to access the private member vars. However I don't think it is a good idea to make them virtual considering that these are private const member variables.

Reply:
Sorry for not being clear. In the set of the protected methods, I think one for the `mVariableDataVector` is missing. Making them virtual is not required.

Reply:
OK! Yes, I already added the missing ones (also for the tolerance vectors).

Reply:
Perfect, thank you.